### PR TITLE
QAT knowledge distillation code example for LLMs

### PR DESCRIPTION
**Pull request by quic-kyunggeu**

### CODE REVIEW

**Review by quic-akhobare — APPROVED** [review]

**Comment by yuyangxie96** [comment]
This training script cannot be used for training because it defaults to fp16, which will cause loss scale overflow and force exit after a few training iterations.
However, this code does not support bf16 either, and I received an error that the quantsim interface does not support bf16. And if it is forcibly changed to fp32, the training will cause a Nan error. Please provide a correct script to start training, thank you.

**Comment by quic-kyunggeu** [comment]
Hi @yuyangxie96, sorry for the inconvenience. Bfloat16 support will be added in https://github.com/quic/aimet/pull/3265.

Besides, just to better understand your situation, are you using some other model that isn't listed in finetune_llm_qat_kd.py?
I personally tested that these models don't end up with scale overflow error (at least for reasonable number of training iterations), but it can easily happen if you are using a different model or different dataset that wasn't explicitly tested. (This is the unfortunate reality of float16 due to its very nature... 😭)
https://github.com/quic/aimet/blob/15ce28726a097ff88fdb4880856ea16bbd3b1dfe/Examples/torch/quantization/llm_qat_kd/finetune_llm_qat_kd.py#L80-L81

In any case, once #3265 gets merged, you'll be able to use bfloat16 from the next release of AIMET. Thanks a lot for raising this issue!

**Comment by yuyangxie96** [comment]
Thank you for your response. I indeed utilized our own model and data. This code fix is quite meaningful, thank you.

**Comment by yuyangxie96** [comment]
@quic-kyunggeu Hello, may I trouble you once more? I have made the changes outlined in https://github.com/quic/aimet/pull/3265 to my local aimet_torch code, and bf16 is now operational. However, I'm unsure why the program still displays a final ppl of NaN. I've also extracted the relevant weights and found that all matrix elements have become NaN. Does this indicate that additional code changes are required within aimet to support bf16 operation?
image

**Comment by quic-kyunggeu** [comment]
Hi @yuyangxie96, I just tried bfloat16 with gpt2-medium in this example, and the nan problem didn't occur in this case.

My take is that it's the model characteristics, and you should probably try to retune the hyperparameters.
The nan issue is very common especially in QAT with sub-float32 dtypes (float16 or bfloat16).
One common recommendation for QAT is using lower learning rate than pretraining because quantization parameters tend to be more sensitive to updates.
